### ISSUE TITLE
Add a nice way to get the router API inside a component

### ISSUE THREAD

**Issue by arunoda** [zeit/next.js#2032]
We need an HOC API like this:

```js
import { withRouter } from 'next/router'

const Comp = (props) => ({ props.router.asPath })
export default withRouter()(Comp)
```

**Comment by nwshane** [comment]
@arunoda Not sure if you want my help, but I would be interested in trying to implement this. Want me to give it a go or do you prefer to do it yourself?

**Comment by arunoda** [comment]
I've to delay this for a week. Yeah! Go ahead. I'll fine tune it.
Allow us to edit changes when submitting the PR.
This is the place you need to add the [router to the context](https://github.com/zeit/next.js/blob/dff20eae0137ee121c703733fbb87084a8cd86d3/lib/app.js).

**Comment by sergeybekrin** [comment]
@nwshane @arunoda sorry for breaking in, but I've been doing same thing with monkey-patching (locally). Just sharing whatever I have.

**Comment by nwshane** [comment]
@sergeybekrin Thanks, these files are helpful.
@arunoda Should I make a branch off of master containing this change? Or off of v3-beta?

**Comment by arunoda** [comment]
@nwshane 

* Use the master
* Yes. Use the using-router. But better to change the example that it can be used in the server. Better example would be the showing the active state of a link.
* run `npm run lint -- --fix`

**Comment by nwshane** [comment]
There's the PR. I took your idea of showing the active state of a link :)

I noticed when implementing this that `props.router.asPath` is `undefined` on the server. Can that be fixed to pass down the correct `asPath` value? Otherwise it still doesn't fix my use case described [here](https://github.com/zeit/next.js/issues/2019), because I need `props.router.asPath` on server and client.

**Comment by arunoda** [comment]
@nwshane shall we discuss this on the PR. Could you tell me about the issue there?

**Comment by ayazhussein** [comment]
Any updates? @arunoda @nwshane Thank you in advance!! :D

**Comment by arunoda** [comment]
Guys, sorry for not taking this yet.
We are currently in feature freeze due to 3.0 release.
Once we did it we'll take this in ASAP.

**Comment by chhuang** [comment]
Will this feature be included in 3.0?

**Comment by yhjor1212** [comment]
Waiting

**Comment by i8ramin** [comment]
Any updates on this? been dying for a way to add an "active" class to my links

**Comment by timneutkens** [comment]
@i8ramin it's going to be merged after v3 lands.

**Comment by HaNdTriX** [comment]
@i8ramin this is the current progress: https://github.com/zeit/next.js/pull/2352

**Comment by alexedev** [comment]
should be merged already?

**Comment by fritz-c** [comment]
This is available in 3.2.0 now.

**Comment by timneutkens** [comment]
@fritz-c thanks, closed 👌
